### PR TITLE
Implementing aave handler

### PR DESCRIPTION
**Pull request by Havoc19**

### CODE REVIEW

File: contracts/handler/Aave/AaveAssetHandler.sol [hunk 1]
```diff
@@ -0,0 +1,755 @@
...
+    address,
+    address asset,
+    address onBehalfOf,
+    uint256 borrowAmount
+  ) external pure returns (bytes memory data) {
+    data = abi.encodeWithSelector(
+      bytes4(keccak256("borrow(address,uint256,uint256,uint16,address)")),
+      asset, // Encode the data to borrow the specified amount
+      borrowAmount,
+      2,
```
Comment (langnavina97) [inline contracts/handler/Aave/AaveAssetHandler.sol:83]: What is this 2 for? Can you add a comment please? Also for the 0 below

File: contracts/front-end-helpers/PortfolioCalculations.sol [hunk 2]
```diff
@@ -721,46 +727,65 @@ contract PortfolioCalculations is ExponentialNoError {
     uint256 totalCollateral
   ) internal pure returns (uint256 debtValue, uint256 percentageToRemove) {
     uint256 feeAmount = (_debtRepayAmount * 10 ** 18 * feeUnit) / 10 ** 22;
+    console.log("feeAmount", feeAmount);
```
Comment (langnavina97) [inline contracts/front-end-helpers/PortfolioCalculations.sol:730]: Remove console.log

File: contracts/handler/Aave/AaveAssetHandler.sol [hunk 3]
```diff
@@ -130,40 +135,36 @@ contract AaveAssetHandler is IAssetHandler {
...
-      DataTypes.ReserveDataLegacy memory data = IPool(comptroller)
+      (, , uint currentVariableDebt, , , , , , ) = IPoolDataProvider(
+        DATA_PROVIDER_ADDRESS
+      ).getUserReserveData(assets[i], account);
+      DataTypes.ReserveDataLegacy memory data = IAavePool(comptroller)
         .getReserveData(asset);
-      if (currentATokenBalance > 0) {
-        lendTokens[lendCount++] = data.aTokenAddress; // Add the asset to the lend tokens if there is a balance
-      }
+      // if (currentATokenBalance > 0) {
```
Comment (langnavina97) [inline contracts/handler/Aave/AaveAssetHandler.sol:165]: Remove comments

File: contracts/handler/Aave/AaveAssetHandler.sol [hunk 4]
```diff
@@ -512,6 +518,7 @@ contract AaveAssetHandler is IAssetHandler {
 
       // Loop through the lending tokens to process each one
       for (uint j = 0; j < lendingTokens.length; ) {
+        console.log("sellAmounts for pull from vault", sellAmounts[j]);
```
Comment (langnavina97) [inline contracts/handler/Aave/AaveAssetHandler.sol:521]: Remove console

File: contracts/handler/ExternalSwapHandler/EnsoHandler.sol [hunk 5]
```diff
@@ -22,58 +23,65 @@ contract EnsoHandler is IIntentHandler, ExternalPositionManagement {
...
+
+    // Validate lengths of input arrays for consistency.
+    uint256 tokensLength = tokens.length;
+
+    if (
+      tokensLength != minExpectedOutputAmounts.length ||
+      tokensLength != callDataEnso.length
+    ) revert ErrorLibrary.InvalidLength();
+
+    console.log(
```
Comment (langnavina97) [inline contracts/handler/ExternalSwapHandler/EnsoHandler.sol:56]: remove console

File: contracts/handler/Aave/AaveAssetHandler.sol [hunk 6]
```diff
@@ -568,16 +575,24 @@ contract AaveAssetHandler is IAssetHandler {
...
+
     //Get price for _protocolToken token
     uint _oraclePrice = IAavePriceOracle(PRICE_ORACLE_ADDRESS).getAssetPrice(
       _underlyingToken
     );
+
+    console.log("_oraclePrice in contract", _oraclePrice);
     //Get price for borrow Balance (amount * price)
-    uint _tokenPrice = (borrowBalance * _oraclePrice) / 10 ** 8;
+    uint _tokenPrice = (borrowBalance * _oraclePrice) / 10 ** 18;
```
Comment (langnavina97) [inline contracts/handler/Aave/AaveAssetHandler.sol:594]: Does this only support tokens with 18 decimals?

File: contracts/front-end-helpers/PortfolioCalculations.sol [hunk 7]
```diff
@@ -488,11 +495,9 @@ contract PortfolioCalculations is ExponentialNoError {
     }
 
     uint256 performanceIncrease = _currentPrice - _highWaterMark;
-    uint256 performanceFee = (performanceIncrease *
+    uint256 performanceFee = ((performanceIncrease *
       _totalSupply *
-      _feePercentage) /
-      ONE_ETH_IN_WEI /
-      TOTAL_WEIGHT;
+      _feePercentage) * ONE_ETH_IN_WEI) / TOTAL_WEIGHT;
```
Comment (langnavina97) [inline contracts/front-end-helpers/PortfolioCalculations.sol:500]: Why did you change the performance fee calculation?

File: contracts/front-end-helpers/PortfolioCalculations.sol [hunk 8]
```diff
@@ -690,11 +727,217 @@ contract PortfolioCalculations is ExponentialNoError {
...
+    amounts = new uint256[](tokenAddresses.lendTokens.length);
+
+    //Get borrow balance for _protocolToken
+    address _underlyingToken = IAaveToken(_protocolToken)
+      .UNDERLYING_ASSET_ADDRESS();
+    (, , uint currentVariableDebt, , , , , , ) = IPoolDataProvider(
+      0x7F23D86Ee20D869112572136221e173428DD740B
+    ).getUserReserveData(_underlyingToken, _user);
+
+    console.log("_underlyingToken", _underlyingToken);
```
Comment (langnavina97) [inline contracts/front-end-helpers/PortfolioCalculations.sol:769]: remove comments

File: contracts/front-end-helpers/PortfolioCalculations.sol (hunk 7 above)
Comment (Havoc19) [reply contracts/front-end-helpers/PortfolioCalculations.sol:500]: Ohkk will revert it

File: contracts/handler/Aave/AaveAssetHandler.sol (hunk 6 above)
Comment (Havoc19) [reply contracts/handler/Aave/AaveAssetHandler.sol:594]: It supports all type of tokens, but converts to 18 decimal

Comment (Havoc19) [reply contracts/handler/Aave/AaveAssetHandler.sol:594]: But thanks for pointing out, I guess there is an issue here, not all aave lending tokens are 18 decimal

Comment (Havoc19) [reply contracts/handler/Aave/AaveAssetHandler.sol:594]: It's correct checked it and our collateral amount to sell is correct

Comment (langnavina97) [reply contracts/handler/Aave/AaveAssetHandler.sol:586]: For any kind of token?

**Review by langnavina97 — APPROVED** [review]

**Review by aj07 — APPROVED** [review]
